File: frontend/onied/src/components/blocks/video/vkVideoProvider.js
import VideoProvider from "./videoProvider";

/**
 * VK Video Provider.
 *
 * @class VkVideoProvider
 * @extends {VideoProvider}
 */
export default class VkVideoProvider extends VideoProvider {
  regex =
    /^(?:https:\/\/)?(?:www\.)?(?:vk\.com|m\.vk\.com)(?:\/video)(?<videoOid>[\d-]+)_(?<videoId>[\d]+)$/;

  getLink(href) {
    const matches = href.match(this.regex);

    return `https://vk.com/video_ext.php?oid=${matches.groups.videoOid}&id=${matches.groups.videoId}`;
  }
}

File: mobile/design_course_blocks/lib/shared/video/providers/vk_video_provider.dart
import 'package:design_course_blocks/shared/video/providers/video_provider.dart';

class VkVideoProvider extends VideoProvider {
  @override
  RegExp get regex => RegExp(
      r'^(?:https:\/\/)?(?:www\.)?(?:vk\.com|m\.vk\.com|vkvideo\.ru)(?:\/video|video-)(?<videoOid>[\d-]+)_(?<videoId>[\d]+)$');

  @override
  String getLink(String href) {
    final matches = regex.firstMatch(href);
    if (matches == null ||
        matches.namedGroup('videoOid') == null ||
        matches.namedGroup('videoId') == null) {
      throw Exception('Невалидный URL VK');
    }

    final videoOid = matches.namedGroup('videoOid')!;
    final videoId = matches.namedGroup('videoId')!;
    return 'https://vk.com/video_ext.php?oid=$videoOid&id=$videoId';
  }
}
